src/test/java/com/amazon/dlic/auth/http/jwt/keybyoidc/SelfRefreshingKeySetTest.java

```diff
@@ -11,108 +11,103 @@
 
-    @Test
-    public void basicTest() throws AuthenticatorUnavailableException, BadCredentialsException {
-        SelfRefreshingKeySet selfRefreshingKeySet = new SelfRefreshingKeySet(new MockKeySetProvider());
+	private SelfRefreshingKeySet selfRefreshingKeySet;
 
-        JsonWebKey key1 = selfRefreshingKeySet.getKey("kid/a");
-        Assert.assertEquals(TestJwk.OCT_1_K, key1.getProperty("k"));
-        Assert.assertEquals(1, selfRefreshingKeySet.getRefreshCount());
+	@Before
+	public void setUp() {
+		selfRefreshingKeySet = new SelfRefreshingKeySet(new MockKeySetProvider());
+	}
 
-        JsonWebKey key2 = selfRefreshingKeySet.getKey("kid/b");
-        Assert.assertEquals(TestJwk.OCT_2_K, key2.getProperty("k"));
-        Assert.assertEquals(1, selfRefreshingKeySet.getRefreshCount());
+	@Test
+	public void getKey_withKidShouldReturnValidKey() throws AuthenticatorUnavailableException, BadCredentialsException {
 
-        try {
-            selfRefreshingKeySet.getKey("kid/X");
-            Assert.fail("Expected a BadCredentialsException");
-        } catch (BadCredentialsException e) {
-            Assert.assertEquals(2, selfRefreshingKeySet.getRefreshCount());
-        }
+		JsonWebKey jsonWebKey = selfRefreshingKeySet.getKey("kid/a");
+		Assert.assertEquals(TestJwk.OCT_1_K, jsonWebKey.getKeyProperty("k"));
+	}
 
-    }
+	@Test
+	public void getKey_withNullKidShouldThrowAuthenticatorUnavailableException()
+			throws AuthenticatorUnavailableException, BadCredentialsException {
 
-    @Test(timeout = 10000)
-    public void twoThreadedTest() throws Exception {
-        BlockingMockKeySetProvider provider = new BlockingMockKeySetProvider();
+		Assert.assertThrows(AuthenticatorUnavailableException.class, () -> selfRefreshingKeySet.getKey(null));
 
-        final SelfRefreshingKeySet selfRefreshingKeySet = new SelfRefreshingKeySet(provider);
+	}
 
-        ExecutorService executorService = Executors.newCachedThreadPool();
+	@Test
+	public void getKeyAfterRefresh_withKidShouldReturnKey()
+			throws AuthenticatorUnavailableException, BadCredentialsException {
 
-        Future<JsonWebKey> f1 = executorService.submit(() -> selfRefreshingKeySet.getKey("kid/a"));
+		JsonWebKey jsonWebKey = selfRefreshingKeySet.getKeyAfterRefresh("kid/b");
+		Assert.assertEquals(TestJwk.OCT_2_K, jsonWebKey.getKeyProperty("k"));
+	}
 
-        provider.waitForCalled();
+	@Test
+	public void getKey_withInvalidDataShouldReturnBadCredentialException()
+			throws AuthenticatorUnavailableException, BadCredentialsException {
 
-        Future<JsonWebKey> f2 = executorService.submit(() -> selfRefreshingKeySet.getKey("kid/b"));
+		Assert.assertThrows(BadCredentialsException.class, () -> selfRefreshingKeySet.getKey("kid/X"));
+	}
 
-        while (selfRefreshingKeySet.getQueuedGetCount() == 0) {
-            Thread.sleep(10);
-        }
+	@Test
+	public void getKeyAfterRefresh_queuedGetCountVariableShouldBeZeroWhenFinishWithAllKeyRefreshes()
```

Comment:
Nice! Do you know if this class has tests that validate scenarios where key refresh fails?

Looks like OIDC Authenticator might use a SelfRefreshingKeySet to retrieve keys from the well_known endpoint (`.../well-known/openid-configuration`) which can fail.

Reply:
I created tests to validate the exceptions that the methods throw when they get null or invalid key, but I'm not sure if this is enough:
- getKey_withNullKidShouldThrowAuthenticatorUnavailableException
- getKey_withInvalidDataShouldReturnBadCredentialException
- getKeyAfterRefresh_withNullKidShouldThrowBadCredentialsException
- getKeyAfterRefresh_withInvalidDataShouldReturnBadCredential

@cwperks  do you think I should cover other scenarios?

Reply:
I think that covers a wide range of scenarios. Nice work on these tests!

Are you still working on this PR or can it be flipped to in review?